File: app/src/main/java/message/smsapp/message/Conversation.java
package message.smsapp.message;

import android.util.Log;

import java.util.ArrayList;

import message.smsapp.activities.MainActivity;

public class Conversation {
	private String phoneNumber;
	private ArrayList<String> messages;
	
	public Conversation(String phoneNumber, ArrayList<String> messages) {
		this.phoneNumber = phoneNumber;
		this.messages = messages;
	}
	
	public static ArrayList<Conversation> generateConversation(ArrayList<String> adress,
	                                                           ArrayList<String> smsContent) {
		ArrayList<Conversation> listOfConversations = new ArrayList<>();
		
		String currentNumber;
		ArrayList<String> messages;
		
		// Add first conversation
		currentNumber = adress.get(0);
		messages = new ArrayList<>();
		
		for (int i = 0; i < smsContent.size(); i++) {
			if (adress.get(i).equals(currentNumber)) {
				messages.add(smsContent.get(i));
			}
		}
		
		listOfConversations.add(new Conversation(currentNumber, messages));
		
		// Add other conversations
		while (!currentNumber.equals("")) {
			currentNumber = Conversation.selectANewConversation(listOfConversations, adress);
			
			for (int i = 0; i < smsContent.size(); i++) {
				if (adress.get(i).equals(currentNumber)) {
					messages.add(smsContent.get(i));
				}
			}
			
			listOfConversations.add(new Conversation(currentNumber, messages));
		}
		
		return listOfConversations;
	}
	
	private static String selectANewConversation(ArrayList<Conversation> list, ArrayList<String>
			adress) {
		ArrayList<String> phoneUsed = new ArrayList<>();
		
		for (int i = 0; i < list.size(); i++) {
			phoneUsed.add(list.get(i).getPhoneNumber());
		}
		
		for (int i = 0; i < adress.size(); i++) {
			if (!phoneUsed.contains(adress.get(i))) {
				return adress.get(i);
			}
		}
		
		return "";
	}
	
	public String getPhoneNumber() {
		return phoneNumber;
	}
	
	public ArrayList<String> getMessages() {
		return messages;
	}
}

File: app/src/main/java/message/smsapp/message/Conversation.kt
package message.smsapp.message

import java.util.*

class Conversation(val phoneNumber: String, val messages: ArrayList<String>) {
    companion object {
		fun generateConversation(adress: ArrayList<String>,
								 smsContent: ArrayList<String>): ArrayList<Conversation> {
            val listOfConversations = ArrayList<Conversation>()
            var currentNumber: String

            // Add first conversation
            currentNumber = adress[0]
            val messages: ArrayList<String> = ArrayList()

            for (i in smsContent.indices) {
                if (adress[i] == currentNumber) {
                    messages.add(smsContent[i])
                }
            }
            listOfConversations.add(Conversation(currentNumber, messages))

            // Add other conversations
            while (currentNumber != "") {
                currentNumber = selectANewConversation(listOfConversations, adress)
                for (i in smsContent.indices) {
                    if (adress[i] == currentNumber) {
                        messages.add(smsContent[i])
                    }
                }
                listOfConversations.add(Conversation(currentNumber, messages))
            }
            return listOfConversations
        }

        private fun selectANewConversation(list: ArrayList<Conversation>, adress: ArrayList<String>): String {
            val phoneUsed = ArrayList<String>()
            for (i in list.indices) {
                phoneUsed.add(list[i].phoneNumber)
            }
            for (i in adress.indices) {
                if (!phoneUsed.contains(adress[i])) {
                    return adress[i]
                }
            }
            return ""
        }
    }

}